🚨 Refactor DETR to updated standards

yonigozlan opened the pull request:
# What does this PR do?

This PR aims at refactoring DETR as part of an effort to standardize vision models in the library, in the same vein as https://github.com/huggingface/transformers/pull/41546.
Expect to see much more PRs like this for vision models as we approach v5!

yonigozlan committed:
refactor attention and many other

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
remove return_dict interface

yonigozlan committed:
improve variable names

yonigozlan committed:
use _can_record_outputs and add real support for pixel and queries masks, vision input embeds and query embeds

yonigozlan committed:
split self attention and cross attention

yonigozlan committed:
nits

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -384,207 +379,253 @@ def __init__(self, embedding_dim=256):
+        self.num_attention_heads = num_attention_heads
+        self.attention_dropout = dropout
+        self.head_dim = hidden_size // num_attention_heads
+        if self.head_dim * num_attention_heads != self.hidden_size:
```
Is this restriction specific to the model. You can also have intermediate sizes different from `hidden_size`, it's just that it's common to have them as same

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -384,207 +379,253 @@ def __init__(self, embedding_dim=256):
+                f" {num_attention_heads})."
             )
         self.scaling = self.head_dim**-0.5
+        self.is_causal = False
```
I'm a bit confused as we have a cross-attention version. Just for clarification, we do not have a causal version? Just used to having decoders with encoder-decoder archs

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -606,96 +647,89 @@ def forward(
                 attention_mask=encoder_attention_mask,
-                spatial_position_embeddings=object_queries,
-                output_attentions=output_attentions,
+                position_embeddings=object_queries_position_embeddings,
```
Should we rename consistently here?

vasqu reviewed (commented):
Some initial thoughts, focused on the masks/interface part

yonigozlan commented on tests/models/detr/test_modeling_detr.py:
```diff
@@ -526,6 +533,18 @@ def test_greyscale_images(self):
+    ):
+        if use_attention_mask:
+            self.skipTest(
+                "This test uses attention masks which are not compatible with DETR. Skipping when use_attention_mask is True."
```
It's more that `_test_eager_matches_sdpa_inference` is not adapted to the vision space (+object queries here). It tries to add a "decoder_input_ids" to the inputs, plus the `seqlen` created for the dummy masks were wrong. Seeing as the function is already quite cluttered and difficult to read, I figured trying to add support for vision model directly there would not be ideal. We can either override the tests in this model specifically, or try to have a more general test for vision models. Another option would be to be able to parameterize the tests by providing how to find the correct seqlen and input names.
I would love some help on this!

yonigozlan replied:
Yes I think it will increase too with us fixing the attention masks for vision models, so we definitely need to improve the base test

yonigozlan committed:
standardize mask handling

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -384,207 +379,239 @@ def __init__(self, embedding_dim=256):
-        position_embedding = DetrLearnedPositionEmbedding(n_steps)
-    else:
-        raise ValueError(f"Not supported {config.position_embedding_type}")
+def eager_attention_forward(
```
Let's use a `# Copied from ...` statement so that during changes we have the connection here

yonigozlan replied:
The goal is to use modular eventually (in this PR even) :)

vasqu replied:
Perfect! Then dont mind this comment :p only want to have some dep here

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -763,6 +781,11 @@ class DetrEncoder(DetrPreTrainedModel):
 
+    _can_record_outputs = {
+        "hidden_states": DetrEncoderLayer,
+        "attentions": OutputRecorder(DetrSelfAttention, layer_name="self_attn", index=1),
```
Do we need the explicit output recorder, iirc `DetrSelfAttention` should work fine in itself

yonigozlan replied:
No indeed I can remove it :)

vasqu reviewed (commented):
Looking good from my side, amazing work! Just left some smaller comments but nothing crazy

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -877,6 +882,12 @@ class DetrDecoder(DetrPreTrainedModel):
+        "hidden_states": DetrDecoderLayer,
+        "attentions": OutputRecorder(DetrSelfAttention, layer_name="self_attn", index=1),
+        "cross_attentions": OutputRecorder(DetrCrossAttention, layer_name="encoder_attn", index=1),
+    }
```
Too many models :D TIL

molbap commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -924,108 +924,68 @@ def forward(
             if self.training:
                 dropout_probability = torch.rand([])
                 if dropout_probability < self.layerdrop:
                     continue
```
not exactly the typical dropout interface, we can maybe take the occasion to update it?

yonigozlan replied:
Yes 😫, I was scared of breaking BC in that case, but maybe it's not so important. It would be great to get rid of non standards dropout elsewhere as well really

molbap commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -1455,104 +1431,84 @@ def forward(
-        # Second, apply 1x1 convolution to reduce the channel dimension to d_model (256 by default)
-        feature_map, mask = features[-1]
-        batch_size, num_channels, height, width = feature_map.shape
+        # Apply 1x1 conv to map (N, C, H, W) -> (N, d_model, H, W), then flatten to (N, HW, d_model)
```
Couple explanatory comments removed/simplified here, no particular opinion on them, what do you think?

molbap reviewed (commented):
Looks niiiice
For the unhappy CI, let's throw the Check Copies away!

yonigozlan commented on src/transformers/modeling_utils.py:
```diff
@@ -206,6 +206,7 @@ def is_local_dist_rank_0():
     "qwen2_5_vl",
     "videollava",
     "vipllava",
+    "detr",
```
Yes that's what made me go crazy haha otherwise `_checkpoint_conversion_mapping` doesn't work.
Note that this is temporary and will be replaced by the new way to convert weights on the fly that @ArthurZucker and @Cyrilvallez are working on.

yonigozlan commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -606,96 +633,89 @@ def forward(
     def __init__(self, config: DetrConfig):
         super().__init__()
-        self.embed_dim = config.d_model
+        self.hidden_size = config.d_model
```
In what way? If users access it directly? In any case I think we really need to standardize these types of variable names, it might be worth slightly breaking BC imo

yonigozlan committed:
update DetrMHAttentionMap

yonigozlan commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -1669,7 +1625,7 @@ def __init__(self, query_dim, hidden_dim, num_heads, dropout=0.0, bias=True, std
+    def forward(self, q, k, mask: Optional[torch.Tensor] = None):
         q = self.q_linear(q)
         k = nn.functional.conv2d(k, self.k_linear.weight.unsqueeze(-1).unsqueeze(-1), self.k_linear.bias)
         queries_per_head = q.view(q.shape[0], q.shape[1], self.num_heads, self.hidden_dim // self.num_heads)
```
Yes I think we could even try to refactor this to use the standard attention module and only take the attention weights! It could be interesting to compare the performance of eager attention vs this implementation (conv2d instead of linear for key proj, and no multiplication by value) vs other attention impl.

yonigozlan committed:
refactor mlp detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
make style

yonigozlan committed:
remve outdated tests which used ".bin" checkpoints

yonigozlan committed:
Updates modeling_detr to newest standards

ArthurZucker commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -1384,19 +1332,18 @@ def __init__(self, config: DetrConfig):
         # segmentation head
         hidden_size, number_of_heads = config.d_model, config.encoder_attention_heads
-        intermediate_channel_sizes = self.detr.model.backbone.conv_encoder.intermediate_channel_sizes
+        intermediate_channel_sizes = self.detr.model.backbone.intermediate_channel_sizes
```
would be nice if this is explicit in the config

ArthurZucker commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -1669,29 +1598,44 @@ def forward(self, x: Tensor, bbox_mask: Tensor, fpns: list[Tensor]):
+        attn_weights = torch.einsum("bqnc,bnchw->bqnhw", query_states * self.scaling, key_states)
+
+        if attention_mask is not None:
+            attn_weights = attn_weights.masked_fill(
```
are we doing this many time?

ArthurZucker reviewed (commented):
looks good in general!

yonigozlan committed:
Review + fix detr weight conversion

yonigozlan committed:
replace einsum, reorder

yonigozlan committed:
refactor rt_detr rt_detr_v2 d_fine to updated library standards

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
fix repo

yonigozlan committed:
Fix test_reverse_loading_mapping test

molbap commented on src/transformers/core_model_loading.py:
```diff
@@ -49,6 +49,38 @@
+    if capturing_group_match:
+        captured_group = capturing_group_match.group(0)
+        pattern = pattern.replace(captured_group, r"\1", 1)
+    return pattern, captured_group
```
For this refactor, it seems unrelated, or is it doing something I'm missing? Can break it off to another PR if independent of DETR standards for commit history (in the target patterns indeed seems to be bugfixing, but we can do in another PR and merge quickly)

molbap commented on src/transformers/core_model_loading.py:
```diff
@@ -460,23 +492,22 @@ def __post_init__(self):
+        capturing_groups_index = 0
         for i, pattern in enumerate(self.source_patterns):
             if r"\1" in pattern:
-                pattern = pattern.replace(r"\1", r"(.+)")
```
indeed this replace seems wrong! cc @Cyrilvallez if you know more

molbap commented on src/transformers/models/d_fine/modular_d_fine.py:
```diff
@@ -508,114 +604,272 @@ def forward(
+        super().__init__(config)
+        self.mlp = DFineMLP(
+            self.hidden_size, config.encoder_ffn_dim, self.hidden_size, 2, config.encoder_activation_function
+        )
```
that's the main new one right, with the `DecoderLayer` too? cleaner

molbap commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +746,155 @@ def forward(
-            outputs += (self_attn_weights, cross_attn_weights)
 
-        return outputs
+class DetrMHAttentionMap(nn.Module):
```
I remember that :O it's better as it is now, but we would really need to use the `attention_interface` in some way here. if not done here I'd be happy to have an Issue opened to do it

molbap reviewed (commented):
Thanks @yonigozlan for the push :pray: we're close to merging this I think, just a few questions / comments

yonigozlan commented on src/transformers/safetensors_conversion.py:
```diff
@@ -28,7 +28,8 @@ def spawn_conversion(token: str, private: bool, model_id: str):
         for line in _sse_connection.iter_lines():
-            line = line.decode()
+            if not isinstance(line, str):
+                line = line.decode()
```
Ah yes I can try to remove this, maybe it's not an issue anymore. Thanks for the reminder 😁

yonigozlan committed:
use modular for RT-DETR

yonigozlan committed:
refactor conditional and deformable detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
use modular for deformable_detr

yonigozlan committed:
use modular for conditional_detr

yonigozlan committed:
fix repo

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +744,154 @@ def forward(
+        x = cur_fpn + nn.functional.interpolate(x, size=cur_fpn.shape[-2:], mode="nearest")
+        x = self.lay5(x)
+        x = self.gn5(x)
+        x = nn.functional.relu(x)
```
Also instead of using relu directly, we can move it to a more general act fn, might be reusable in the future then

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +744,154 @@ def forward(
+
+        query = (
+            (query_states * self.scaling)
+            .permute(0, 2, 1, 3)
```
`transpose(1,2)` is easier to read

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +744,154 @@ def forward(
+
+        attn_weights = (
+            torch.matmul(query, key.transpose(1, 2))
+            .view(batch_size, num_heads, num_queries, height, width)
```
We could also define this shape above

vasqu reviewed (commented):
Leaving a review here just in case (only looked at detr) - maybe some stuff is already been said so sorry if that's the case

The comments are mainly for a few more things we could improve it feels like. Looks overall pretty good tho!

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
refactor DetrMaskHeadSmallConv

yonigozlan committed:
fix modular

yonigozlan committed:
Temporarily remove outdated copied from

yonigozlan commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -725,17 +901,25 @@ class DetrPreTrainedModel(PreTrainedModel):
+    _supports_flex_attn = True
+    _keys_to_ignore_on_load_unexpected = [
+        r"detr\.model\.backbone\.model\.layer\d+\.0\.downsample\.1\.num_batches_tracked"
+    ]
```
These were not used before the refactor either, and are causing some annoying warnings when loading the model

yonigozlan committed:
fix consistency

yonigozlan commented:
Hello @vasqu ! Thanks a lot for the reviews, I have addressed your comments and refactored DetrMaskHeadSmallConv.

As this PR is getting quite a big, and we need some of the refactored models merged quite quickly as there a few PRs open depending on them, I removed some outdated copied from that I'll change to modular in a subsequent PR, so that we can merge this PR.

molbap commented on src/transformers/conversion_mapping.py:
```diff
@@ -163,6 +163,55 @@ def _build_checkpoint_conversion_mapping():
+            WeightRenaming("out_proj", "o_proj"),
+            WeightRenaming(r"layers.(\d+).fc1", r"layers.\1.mlp.layers.0"),
+            WeightRenaming(r"layers.(\d+).fc2", r"layers.\1.mlp.layers.1"),
+        ],
```
the differences are subtle but there are some, tried to shepherd them a bit. IMO not that easy to find a base canonical rename here due to base names being slightly different from model to model. You could indeed do something like

```python
mapping["rt_detr"] = mapping["detr"][1:]
```
but the rest would save one line at best :X (and not be extra readable)
```python
  mapping["deformable_detr"] = [                                                                                     
      mapping["detr"][0],                                          
      WeightRenaming("self_attn.out_proj", "self_attn.o_proj"),                                                      
      *mapping["detr"][2:],                                                               
  ]       
```

molbap commented on src/transformers/models/conditional_detr/modular_conditional_detr.py:
```diff
@@ -123,4 +161,808 @@ def post_process_semantic_segmentation(self, outputs, target_sizes: list[tuple[i
+        logits = self.class_labels_classifier(sequence_output)
+
+        reference = outputs.reference_points
+        reference_before_sigmoid = inverse_sigmoid(reference).transpose(0, 1)
```
```suggestion
        reference_points_logits = inverse_sigmoid(reference).transpose(0, 1)
```

molbap commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +752,172 @@ def forward(
+        for fpn_stage, fpn_feat in zip(self.fpn_stages, fpn_features):
+            hidden_states = fpn_stage(hidden_states, fpn_feat)
+
+        return self.output_conv(hidden_states)
```
much cleaner without the abbreviations, thanks! FPN is already not easy to understand hehe

molbap reviewed (approved):
Left additional comments/suggestions to help cross the finish line! Looking good

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +752,172 @@ def forward(
+        key = key_states.permute(0, 1, 3, 4, 2).contiguous().view(key_shape)
+
+        attn_weights = (
+            (torch.matmul(query, key.transpose(1, 2)) * self.scaling).view(attn_weights_shape).transpose(1, 2)
```
```suggestion
            (torch.matmul(query * self.scaling, key.transpose(1, 2))).view(attn_weights_shape).transpose(1, 2)
```
The order may have a small influence

vasqu commented on src/transformers/models/detr/modeling_detr.py:
```diff
@@ -703,19 +752,172 @@ def forward(
+                DetrFPNFusionStage(fpn_channels[1], hidden_size // 4, hidden_size // 8, activation_function),
+                DetrFPNFusionStage(fpn_channels[2], hidden_size // 8, hidden_size // 16, activation_function),
+            ]
+        )
```
So much cleaner :pray:

vasqu commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -1149,72 +1048,24 @@ def __init__(self, config: RTDetrConfig):
 
         hidden_states = inputs_embeds
 
         encoder_states = () if output_hidden_states else None
```
Hmm, can we not add this to `_can_record_outputs`

yonigozlan replied:
Fixed!

vasqu reviewed (commented):
Talked with @molbap internally and I think we agree that it doesn't make sense to force this merge just to split refactoring again. Let's aim for quality in this refactor

We will probably merge the model PR as is and add this to this refactor after merge. Otherwise, we will suffer on both sides - crunch time on the model PR and less quality on the refactor (e.g. another set of TODOs)

I've added a few smaller comments meanwhile

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
Improve DetrMHAttentionMap

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
Fix torch functional import aliases

yonigozlan commented on src/transformers/models/conditional_detr/modular_conditional_detr.py:
```diff
@@ -1,15 +1,53 @@
+import math
```
Indeed thanks for the heads up!

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
fix after merge with main

yonigozlan committed:
Fix missing copyrights

yonigozlan committed:
Refactor HybridEncoder rt_detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan commented on src/transformers/models/rt_detr/modeling_rt_detr.py:
```diff
@@ -735,7 +689,7 @@ def forward(
             batch_size, num_queries, self.n_heads, self.n_levels * self.n_points
         )
-        attention_weights = F.softmax(attention_weights, -1).view(
+        attention_weights = softmax(attention_weights, -1).view(
```
Fixed :)

yonigozlan committed:
tie weights fix deformable detr

yonigozlan committed:
Merge branch 'main' into refactor-detr

yonigozlan committed:
Merge branch 'main' into refactor-detr

yonigozlan committed:
Merge branch 'main' into refactor-detr

yonigozlan committed:
Merge remote-tracking branch 'upstream/main' into refactor-detr

yonigozlan committed:
Refactor pp docs layout + fix fp16 overflow

yonigozlan committed:
fix modular

yonigozlan committed:
fix modular

yonigozlan commented on src/transformers/models/pp_doclayout_v3/modular_pp_doclayout_v3.py:
```diff
@@ -713,7 +715,7 @@ def mask_to_box_coordinate(mask, dtype):
     x_max = x_coords_masked.flatten(start_dim=-2).max(dim=-1).values + 1
     x_min = (
-        torch.where(mask, x_coords_masked, torch.tensor(1e8, device=mask.device, dtype=dtype))
+        torch.where(mask, x_coords_masked, torch.tensor(torch.finfo(dtype).max))
```
Note: This was causing overflow issues in float16
Cc @zhang-prog @molbap

yonigozlan committed:
fix deformable detr tests

yonigozlan committed:
Merge branch 'main' into refactor-detr
